Add union-type support

KavinduZoysa opened the pull request:
## Purpose
$subject

Fixes #112 

## Goals
Adding union-type support.

## Approach
The approach is discussed in https://github.com/ballerina-platform/nballerina/discussions/113.
As a summary,
if we have `int|string a = 5;`, we create a struct as follows and assign it to `a`
```
{
   inherentType : int, (should be mangled name)
   valueRef : reference to 5
}
```

## Automation tests
 - Unit tests 
   Added unit tests

Kishanthan commented on test/types/union/unionTypeReturn.bal:
```diff
@@ -0,0 +1,13 @@
+int _bal_result = 0;
+
+function foo() returns int|string {
+  return 221;
```
Please fix the formatting issues in all the bal file here.

KavinduZoysa replied:
Fixed.

ruvi-d reviewed (changes requested):
There are some issues with the codegen when there are both `any` and `union` together.

Please verify with the following two examples:
```
int _bal_result = 0;
public function main() 
{
    any var1 = 42;
    int result1 = <int>var1;
    int | string var2 = 7;
    int result2 = <int>var2;
    _bal_result = result2;
}
```
&
```
int _bal_result = 0;
public function main() 
{
    int | string var1 = 42;
    any result1 = <any>var1;
    _bal_result = <int>result1;
}
```

KavinduZoysa commented on compiler/llvmTypeCastInsn.cpp:
```diff
@@ -20,75 +20,53 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
-      // lhsType or not by calling a runtime library function in Rust.
-      // And the return value of the call should be checked.
-      //  If it returns false, then should be a call to abort().
+  rhsOpRef = funcObj->getLocalVarRefUsingId(rhsOpName);
+  lhsOpRef = funcObj->getLocalVarRefUsingId(lhsOpName);
```
ACK. will check and get back

KavinduZoysa replied:
Fixed this.

ruvi-d reviewed (changes requested):
Shall we add a test case to cover set and get from Global Any and Union type variables?

ayonam commented on compiler/llvmTypeCastInsn.cpp:
```diff
@@ -17,78 +17,71 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
-      LLVMBuildBitCast(builder, rhsOpRef, lhsTypeRef, "data_cast");
+    if (lhsTypeTag == TYPE_TAG_UNION || lhsTypeTag == TYPE_TAG_ANY) {
+        LLVMValueRef rhsVarOpRef = getFunction()->getLocalToTempVar(rhsOp);
+        LLVMBuildStore(builder, rhsVarOpRef, lhsOpRef);
```
How can you do this store if it is a TYPE_TAG_UNION?  You need to perform a runtime check if the last assigned type of the RHS is a part of the UNION that the LHS holds.

KavinduZoysa replied:
In this, if block (L:52), we know both lhs and rhs parts are either union or any. That's why I simply assign the struct to lhs. If we have to do any runtime check we can do it on #147

ayonam commented on compiler/llvmTypeCastInsn.cpp:
```diff
@@ -17,78 +17,71 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
-    } else {
-      LLVMBuildBitCast(builder, rhsOpRef, lhsTypeRef, "data_cast");
+    if (lhsTypeTag == TYPE_TAG_UNION || lhsTypeTag == TYPE_TAG_ANY) {
+        LLVMValueRef rhsVarOpRef = getFunction()->getLocalToTempVar(rhsOp);
```
Why are we doing this here?  Haven't we already got the RHS OpRef somewhere above?  And if we want to do this here, then if RHS Op is not a local, what do we do?

KavinduZoysa replied:
Here I have used the `getLocalToTempVar` method to create a temp variable from rhs. If rhsOp is global, it is handling Inside the `getLocalToTempVar` method

ayonam commented on compiler/llvmTypeCastInsn.cpp:
```diff
@@ -17,78 +17,71 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
+    LLVMValueRef dataLoad = LLVMBuildLoad(builder, data, "");
+    LLVMValueRef castResult =
+        LLVMBuildBitCast(builder, dataLoad, lhsTypeRef, lhsOpName.c_str());
+    LLVMValueRef castLoad = LLVMBuildLoad(builder, castResult, "");
```
Why do you need this load here?  The Cast Result should already contain the result of the cast operation and you should be able to store that into the LHS Op.

KavinduZoysa replied:
To get the value we need to do a load here.
```
  %"%43" = bitcast i8* %1 to i32*
  %2 = load i32, i32* %"%43", align 4
  store i32 %2, i32* %"%4", align 4
```

ayonam commented on compiler/llvmTypeCastInsn.cpp:
```diff
@@ -17,78 +17,71 @@ void TypeCastInsn::translate(LLVMModuleRef &modRef) {
+    LLVMBuildStore(builder, castLoad, lhsOpRef);
+  } else if (lhsTypeTag == TYPE_TAG_UNION || lhsTypeTag == TYPE_TAG_ANY) {
+    LLVMValueRef structAllocaRef =
+        funcObj->getLocalVarRefUsingId(getLhsOperand()->name());
```
Again why are we doing this here?  The code at lines 32-38 is already getting the LHS Op Ref, isn't it?  Here, if the LHS is a global variable, the the call to getLocalVarRefUsingId() could return NULL.

KavinduZoysa replied:
Fixed.

ruvi-d commented on compiler/llvmFunctionGen.cpp:
```diff
@@ -91,6 +91,9 @@ LLVMTypeRef BIRFunction::getLLVMFuncRetTypeRefOfType(VarDecl *vDecl) {
     return LLVMPointerType(LLVMInt8Type(), 0);
+  case TYPE_TAG_ANY:
+  case TYPE_TAG_UNION:
+    return wrap(getPkgAddress()->getStructType());
```
we will address that under #144

ayonam replied:
Fine.  Thanks.

ruvi-d commented on compiler/llvmFunctionGen.cpp:
```diff
@@ -106,12 +109,9 @@ void BIRFunction::translateFunctionBody(LLVMModuleRef &modRef) {
-    localVarRefs.insert({locVar->getVarName(), localVarRef});
+    string localVarName = locVar->getVarName();
+    LLVMValueRef localVarRef = LLVMBuildAlloca(builder, varType, localVarName.c_str());
+    localVarRefs.insert({localVarName, localVarRef});
```
The call to `getLLVMTypeRefOfType()` handles ANY type and returns the correct type to the smart pointer type.

KavinduZoysa replied:
Yes, as @ruvi-d mentioned, I have done a refactoring here, still, the functionality is same.

ruvi-d commented on test/casting/anyTypeReturn.bal:
```diff
@@ -0,0 +1,13 @@
+// RUN: JAVA_HOME=%java_path %testRunScript %s %nballerinacc | filecheck %s
```
I believe that we have have to wait till #132 (mangled type string support) to support `is type` checks and failing tests.

ayonam replied:
Okay.  That's fine.  But then can we please add this requirement as a new issue and mark #132 as a dependency for the same?

ayonam replied:
@ruvi-d Have you marked this dependency on #132 ?  If yes, then please resolve this conversation.

KavinduZoysa replied:
created an issue to track this https://github.com/ballerina-platform/nballerina/issues/152

KavinduZoysa commented on compiler/llvmPackageGen.cpp:
```diff
@@ -78,12 +77,8 @@ void BIRPackage::translate(LLVMModuleRef &modRef) {
-        paramTypes[i] = wrap(structType);
-      } else {
-        paramTypes[i] = birFunc->getLLVMTypeRefOfType(
+      paramTypes[i] = birFunc->getLLVMTypeRefOfType(
```
What we discussed was, to assign the struct type only for the union types variables. Since we are already doing the same thing for any type, with this change, we assign struct type for both union and any. If you look at the `getLLVMTypeRefOfType()` method, you will see I have handled any and union types together.

Do we need to assign struct type for simple types such as int, string, bool etc? I think we did not discussed such a thing.

KavinduZoysa commented on test/union/simpleUnionAssignment3.bal:
```diff
@@ -0,0 +1,9 @@
+
+public function main() {
+    int|string u = "test string";
+    string s = <string>u;
```
@ruvi-d, can we check value assigned for string variable?

KavinduZoysa replied:
@ruvi-d, can we check value assigned for string variable?

KavinduZoysa replied:
or we can do somthing like this right? `_bal_result = s == "test string"`

ruvi-d replied:
not yet.. we need #120. We can revamp the tests when it get merged.

ayonam replied:
@ruvi-d @KavinduZoysa 
Please mark this requirement against #120 and then resolve this conversation.  If it requires me to resolve this, then let me know.  I will do the needful.

KavinduZoysa replied:
created an issue to track this https://github.com/ballerina-platform/nballerina/issues/152

KavinduZoysa commented on test/union/simpleUnionAssignment4.bal:
```diff
@@ -0,0 +1,12 @@
+
+public function main() {
+    any var1 = 42;
+    int result1 = <int>var1;
```
changed the test case

ruvi-d commented:
> Shall we add a test case to cover set and get from Global Any and Union type variables?

Reminder to add a test case for Global Any and Union type variables

KavinduZoysa commented:
> > Shall we add a test case to cover set and get from Global Any and Union type variables?
> 
> Reminder to add a test case for Global Any and Union type variables

I checked it, I think we have to wait until we fix this issue https://github.com/ballerina-platform/nballerina/issues/149, because union/any types are initialized inside `.<init>` function. 

```
public ..<init> function() -> error{map<anydata | readonly>} | () {
    %0(RETURN) error | ();
    %0(RETURN) error | ();
    %1(TEMP) typeDesc<any | error>;
    %5(TEMP) int;

    bb0 {
;
        $annotation_data = NewMap %1;
        _bal_result = ConstLoad 20;
        %5 = ConstLoad 7;
        u = <int | string> %5;
        %0 = ConstLoad 0;
        GOTO bb2;
    }
    bb1 {
        GOTO bb2;
    }
    bb2 {
        return;
    }
}
```


If we generate the .ll file, struct is not initialized.

`@u = global %struct.smtPtr zeroinitializer, align 4`

KavinduZoysa commented:
> > Shall we add a test case to cover set and get from Global Any and Union type variables?
> 
> Reminder to add a test case for Global Any and Union type variables

Added tests related to global vars

ruvi-d commented on test/casting/any_global.bal:
```diff
@@ -0,0 +1,11 @@
+    int x = <int>a;
+    _bal_result = x;
+}
+// CHECK: RETVAL=23
```
No new line at end of file

KavinduZoysa replied:
Added

ruvi-d reviewed (approved):
Good to merge. Just one test file missing an EoF

KavinduZoysa commented:
@ayonam, could you please check this PR? I have addressed the suggestions.

ayonam commented:
> @ayonam, could you please check this PR? I have addressed the suggestions.

@KavinduZoysa I could figure out one of the comments above which I could resolve.  But for the rest I'm thoroughly confused.  Sorry about that.  Old age is catching up :-)

Will it be possible for you to open a fresh PR with your latest changes please?  I can review it then and you can merge that one instead.  Thanks.

KavinduZoysa commented:
> > @ayonam, could you please check this PR? I have addressed the suggestions.
> 
> @KavinduZoysa I could figure out one of the comments above which I could resolve. But for the rest I'm thoroughly confused. Sorry about that. Old age is catching up :-)
> 
> Will it be possible for you to open a fresh PR with your latest changes please? I can review it then and you can merge that one instead. Thanks.

@ayonam, this is the new PR https://github.com/ballerina-platform/nballerina/pull/167

ayonam commented:
> > > @ayonam, could you please check this PR? I have addressed the suggestions.
> > 
> > 
> > @KavinduZoysa I could figure out one of the comments above which I could resolve. But for the rest I'm thoroughly confused. Sorry about that. Old age is catching up :-)
> > Will it be possible for you to open a fresh PR with your latest changes please? I can review it then and you can merge that one instead. Thanks.
> 
> @ayonam, this is the new PR #167

Thanks @KavinduZoysa.  I will review 167 and give my comments.
